### ISSUE TITLE
Open Grant Proposal: Filecoin Lotus VS Code Extension

### ISSUE THREAD

**Issue by yjliu-dataseek** [filecoin-project/devgrants#1198]
# Open Grant Proposal: `Filecoin Lotus VS Code Extension`

**Name of Project:**
Filecoin Litenode VS Code Extension

**Proposal Category:**
`integration-adoption`

**Proposer:**
liubosdzl33

**(Optional) Technical Sponsor:** 
None

**Do you agree to open source all work you do on behalf of this RFP and dual-licence under MIT, APACHE2, or GPL licences?:** 
Yes

# Project Description

#### The Problem with Working with Filecoin Testnet
To first time blockchain developers, setting up a Filecoin Lotus node to interact with Filecoin chain in testnet can be challenging. 

Filecoin Lotus VS Code extension intends to provide an easy way for developers to run a Filecoin node within the VS Code IDE environment much quicker than installing via CLI tools step by step.

The extension should allow developers to store data, retrieve data, interact with wallet, get API token and access Lotus API.

There are two parts to this project: 

_***Part 1***: establish an automation of generating binary of Lotus for Linux & Mac_

This step is likely to be a Github action automation that builds to the Lotus release branch so the latest stable version of Lotus can always be made available for new users of the VS code extension. Testnet/calibration net only. 

_***Part 2***: develop the VS Code extension to download binary and install on user’s VS code global storage and run daemon in the background_

The VS code extension will also download the latest snapshot of calibration net and start the daemon for syncing with the block.

#### What is out of scope? 
- The project will not run a miner module or worker module
- The project is not intended for storage providing
- The project is not making a local devnet, mainnet or FVM testnet
- The project will not create binary code that works for Windows

## Value
Successful completion of the project will bring Filecoin testnet developer experience a lot better, allowing developers to interact with the chain quickly and reduce the frustration of the environment issues that often happens. 

The major risks in the project: 
- Unable to automate CD/CI pipeline for making Lotus binary for targeted OSs
- VS code environment unable to allow Lotus daemon to run
- Major breaking changes to Lotus code that does not allow it to complete

## Deliverables

The final product should be a “Filecoin Lotus VS Code Extension” listed on VS Code Extension market that allows developers to install and run . 

A full list of features: 
- Able to verify successful installation of Lotus full node (calibration net)
- Start Lotus daemon
- Get Lotus API Token
- Access Lotus JSON RPC APIs
- Interact with lotus for wallet management via CLI
- Display sync status
- Allow users to send Store Data deals (smaller than 32 GB, sector size)
- Allow user to send Retrieve Data deals

#### --- Milestone 1 ---

Design UI, detailed breakdown of technical work, build the binary github action CD/CI pipeline. 

Tasks: 
- Design UI component of the VS code
- Work out internal workflows of binary building
- Build a Github repo that fetch latest stable version of Lotus 
- Setup Github action workflow to generate binary for Linux & MacOS 

This repo can be used by the community and other projects to provide faster testnet Lotus installation.

Time: 6 weeks
Budget: USD $25,000 

#### --- Milestone 2 ---

Create VS Code Extension project. Complete core functions of the extension.

Tasks: 
- Able to verify successful installation of Lotus full node (calibration net)
- Start Lotus daemon
- Get Lotus API Token
- Access Lotus JSON RPC APIs

Time: 8 weeks
Budget: USD $35,000 

#### --- Milestone 3 ---

Complete rest of the functions of VS Code.

Tasks: 
- Interact with lotus for wallet management via CLI
- Display sync status
- Allow users to send  Store Data deals (smaller than 32 GB, sector size)
- Allow user to send Retrieve Data deals

Time: 6 weeks
Budget: USD $25,000 

#### --- Milestone 4 ---

Final bug fixes, improvements and release to VS Code Extension store

Tasks: 
- End to end testing
- User testing
- Bug fixes discovered in the testing
- Small improvement implementation
- Go through release process of VS code

Time: 4 weeks
Budget: USD $10,000 

## Total Budget Requested
Total budget USD $95,000

## Maintenance and Upgrade Plans

The build pipeline should take care of future updates automatically. We will constantly come back to fix issues caused by upgrades. We will also welcome community contribution of the project. 

# Team

## Team Members

[YJ Liu](https://github.com/yjliu-dataseek) - Project Manager - yjliu@dataseek.info

[Bo Liu](https://github.com/liubosdzl33) - Lead Developer - liubo@dataseek.info

[Stan Young](https://github.com/user3301) - DevOps Engineer - mokuo@dataseek.info

## Team Website

https://dataseek.info/

## Relevant Experience

Previously received grant #775.

**Comment by ErinOCon** [comment]
Hi @yjliu-dataseek, thank you for your proposal! We can consider this project for a grant once the active grant on file is complete. Please feel welcome to reapply once the final milestone has been completed and reviewed.

**Comment by yjliu-dataseek** [comment]
Hi @ErinOCon, can we please revisit this proposal since our last grant has been completed. Thank you.
